### PR TITLE
Address Travis spurious failures

### PR DESCRIPTION
**Pull request by theuni**
As discussed in last week's meeting. The failures are annoying.

I'm sure that this isn't the best approach, but these rpcs seem generally useful, and I believe this fixes the current race issue.

The problem is that the wallet is not yet finished dealing with transactions before getblockcount/getbestblockhash return new values, so assertions about balances are racy. The approach I've taken here is to wait for a signal that the processing is complete, then cache the new height/hash on the rpc side. That allows for blocking in the call rather than polling, and it ensures that we're synced up before it returns. It also means no cs_main locking.

To fix sync_blocks in particular, wait for height 0 from all nodes. That returns the current height/hash instantly, since it should never actually have to wait. Then, grab the current height from all nodes, take the max, and wait on that height for all nodes. Repeat until all heights match.

@sdaftuar Mentioned that it may be useful to a wait_for_wallet() call, which would presumably work similarly, though I'm afraid it may introduce new race issues. Suggestions on that approach are also welcome.

### CODE REVIEW

**Comment by laanwj** [comment]
These RPCs happen to also wait for wallet processing at this point, but that will change if wallet processing becomes asynchronous.

ACK on this as a stopgap measure to prevent the synchronization issues.

But I think there should be a warning in their help to not rely on this behavior "as the tests do".

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -168,6 +180,138 @@ UniValue getbestblockhash(const UniValue& params, bool fHelp)
...
+    ret.push_back(Pair("hash", block.hash.GetHex()));
+    ret.push_back(Pair("height", block.height));
+    return ret;
+}
+
+UniValue waitforblock(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 2)
+        throw runtime_error(
+            "waitforblock\n"
```
Comment (jonasschnelli) [inline src/rpc/blockchain.cpp]: nit: `s/waitforblock/waitforblock <blockhash> <timeout>`
I guess most commands help have it like this.

**Comment by jonasschnelli** [comment]
Concept ACK.
I like this approach. Also, it's a preform of longpoll RPC notifications proposed in https://github.com/bitcoin/bitcoin/pull/7949 which I think are generally useful.

**Comment by laanwj** [comment]
Gah I'm just going to go ahead and merge this, I need the tests to pass, it's low-risk (largest risk is to break the tests even more, well...) and fixups can be done later.
utACK d6a5dc4

File: qa/rpc-tests/test_framework/util.py [hunk 2]
```diff
@@ -125,12 +125,16 @@ def sync_blocks(rpc_connections, wait=1, timeout=60):
...
+    maxheight = 0
     while timeout > 0:
-        tips = [ x.getbestblockhash() for x in rpc_connections ]
+        tips = [ x.waitforblockheight(maxheight, int(wait * 1000)) for x in rpc_connections ]
+        heights = [ x["height"] for x in tips ]
         if tips == [ tips[0] ]*len(tips):
             return True
-        time.sleep(wait)
+        if heights == [ heights[0] ]*len(heights): #heights are the same but hashes are not
+            raise AssertionError("Block sync failed")
```
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: Nit: `mempool_packages.py` relies on this not being thrown.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -168,6 +180,138 @@ UniValue getbestblockhash(const UniValue& params, bool fHelp)
...
+
+UniValue waitfornewblock(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() > 1)
+        throw runtime_error(
+            "waitfornewblock\n"
+            "\nWaits for a specific new block and returns useful info about it.\n"
+            "\nReturns the current block on timeout or exit.\n"
+            "\nArguments:\n"
+            "1. timeout (milliseconds) (int, optional, default=false)\n"
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: `default=0`?

**Comment by MarcoFalke** [comment]
Just to be sure: The new rpc calls are not meant for general consumption but serve as a temporary fix for the tests?

File: qa/rpc-tests/test_framework/util.py (hunk 2 above)
Comment (sdaftuar) [inline qa/rpc-tests/test_framework/util.py]: There's a slightly bigger issue that @morcos just noticed, which is that the block notification signals doesn't quite work with invalidateblock right now.

I think we can work around it by having the invalidateblock rpc handler explicitly call the ui signals in this case (which is probably the right thing to do in general).

**Comment by luke-jr** [comment]
Does this need a backport?

**Comment by MarcoFalke** [comment]
@luke-jr No, the commit that reworked the locks is only in master

**Comment by MarcoFalke** [comment]
This will also break the tests, when an old binary is used for comp. checks.

**Comment by MarcoFalke** [comment]
There were suggestions by @morcos and @TheBlueMatt to address the failures in a less breaking way during one of the last meetings, but I couldn't find any follow ups.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -168,6 +180,138 @@ UniValue getbestblockhash(const UniValue& params, bool fHelp)
...
+    ret.push_back(Pair("hash", block.hash.GetHex()));
+    ret.push_back(Pair("height", block.height));
+    return ret;
+}
+
+UniValue waitforblockheight(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 2)
+        throw runtime_error(
+            "waitforblock\n"
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: waitforblockheight

**Pull request merged**